pkg/downwardmetrics/virtio-serial/server.go

```diff
@@ -0,0 +1,276 @@
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	"kubevirt.io/kubevirt/pkg/downwardmetrics/vhostmd/api"
+	"kubevirt.io/kubevirt/pkg/util"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+
+	diskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
+	metricsScraper "kubevirt.io/kubevirt/pkg/monitoring/domainstats/downwardmetrics"
+)
+
+const connectAttempts = 6
+const standardLauncherSocketPath = "/run/kubevirt/sockets/launcher-sock"
+
+var (
+	invalidRequest = []byte("INVALID REQUEST\n\n")
+	emptyMetrics   = []byte("<metrics><!-- host metrics not available --><!-- VM metrics not available --></metrics>")
+)
+
+func RunDownwardMetricsVirtioServer(nodeName string, stop chan struct{}) error {
+	report, err := newMetricsReporter(nodeName)
+	if err != nil {
+		return err
+	}
+
+	server := downwardMetricsServer{
+		connectAttempts:    connectAttempts,
+		virtioSerialSocket: downwardmetrics.DownwardMetricsChannelSocket,
+		report:             report,
+	}
+	return server.start(stop)
+}
+
+type metricsReporter func() (*api.Metrics, error)
+
+func newMetricsReporter(nodeName string) (metricsReporter, error) {
+	exists, err := diskutils.FileExists(standardLauncherSocketPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.New("virt-launcher socket not found")
+	}
+
+	isolationDetector := isolation.NewSocketBasedIsolationDetector(util.VirtShareDir)
+	scraper := metricsScraper.NewScraper(nodeName, isolationDetector)
+
+	return func() (*api.Metrics, error) {
+		return scraper.Report(standardLauncherSocketPath)
+	}, nil
+}
+
+type downwardMetricsServer struct {
+	connectAttempts    uint
+	virtioSerialSocket string
+	report             metricsReporter
+}
+
+func (s *downwardMetricsServer) start(stop chan struct{}) error {
+	conn, err := connect(s.virtioSerialSocket, s.connectAttempts)
+	if err != nil {
+		return err
+	}
+
+	go s.serve(conn, stop)
+	return nil
+}
+
+func (s *downwardMetricsServer) serve(conn net.Conn, stop chan struct{}) {
+	defer func(conn net.Conn) {
+		err := conn.Close()
+		if err != nil {
+			log.Log.Reason(err).Warning("virtio-serial failed to close the connection")
+		}
+	}(conn)
+
+	type reqResult struct {
+		request string
+		err     error
+	}
+	newRequest := make(chan reqResult)
+	defer close(newRequest)
+	reader := textproto.NewReader(bufio.NewReader(conn))
+
+	for {
+		go func() {
+			// Let's ignore the possible panic for sending on a closed channel
+			defer func() { _ = recover() }()
+
+			requestLine, err := waitForRequest(reader)
+			if err != nil && errors.Is(err, net.ErrClosed) {
+				// Closing the connection signals that the server should stop
+				return
+			}
+
+			newRequest <- reqResult{request: requestLine, err: err}
+		}()
+
+		select {
+		case res := <-newRequest:
+			if res.err != nil {
+				if errors.Is(res.err, io.EOF) {
+					// qemu closed the connection
+					return
+				}
+				log.Log.Reason(res.err).Warning("virtio-serial socket read failed")
```

Comment:
Can the sender handle error messages, i.e. does it make sense to send an error / error code back to the client?